src/linker/Linker/Driver.cs

```diff
@@ -99,49 +99,54 @@ public static bool ProcessResponseFile (string[] args, out Queue<string> result)
 			return true;
 		}
 
-		public static void ParseResponseFileLines (IEnumerable<string> responseFileLines, Queue<string> result)
+		public static void ParseResponseFile (TextReader reader, Queue<string> result)
 		{
-			foreach (var rawResponseFileText in responseFileLines) {
-				var responseFileText = rawResponseFileText.Trim ();
-				int idx = 0;
-				while (idx < responseFileText.Length) {
-					while (idx < responseFileText.Length && char.IsWhiteSpace (responseFileText[idx])) {
-						idx++;
-					}
-					if (idx == responseFileText.Length) {
+			int cur;
+			while ((cur = reader.Read ()) > 0) {
+				// skip whitespace
+				while (char.IsWhiteSpace ((char) cur)) {
+					if ((cur = reader.Read ()) < 0)
 						break;
+				}
+				if (cur < 0) // EOF
+					break;
+
+				StringBuilder argBuilder = new StringBuilder ();
+				bool inquote = false;
+
+				// build up an argument one character at a time
+				while (true) {
+					bool copyChar = true;
+					int numBackslash = 0;
+
+					// count backslashes
+					while (cur == '\\') {
+						numBackslash++;
+						if ((cur = reader.Read ()) < 0)
+							break;
 					}
-					StringBuilder argBuilder = new StringBuilder ();
-					bool inquote = false;
-					while (true) {
-						bool copyChar = true;
-						int numBackslash = 0;
-						while (idx < responseFileText.Length && responseFileText[idx] == '\\') {
-							numBackslash++;
-							idx++;
-						}
-						if (idx < responseFileText.Length && responseFileText[idx] == '"') {
-							if ((numBackslash % 2) == 0) {
-								if (inquote && (idx + 1) < responseFileText.Length && responseFileText[idx + 1] == '"') {
-									idx++;
-								} else {
-									copyChar = false;
-									inquote = !inquote;
-								}
+					if (cur == '"') {
+						if ((numBackslash % 2) == 0) {
+							if (inquote && (reader.Peek () == '"')) {
+								// handle "" escape sequence in a quote
+								cur = reader.Read ();
+							} else {
+								// unescaped " begins/ends a quote
+								copyChar = false;
+								inquote = !inquote;
 							}
-							numBackslash /= 2;
-						}
-						argBuilder.Append (new String ('\\', numBackslash));
-						if (idx == responseFileText.Length || (!inquote && Char.IsWhiteSpace (responseFileText[idx]))) {
-							break;
 						}
-						if (copyChar) {
-							argBuilder.Append (responseFileText[idx]);
-						}
-						idx++;
+						// treat backslashes before " as escapes
+						numBackslash /= 2;
 					}
-					result.Enqueue (argBuilder.ToString ());
+					argBuilder.Append (new String ('\\', numBackslash));
```

Comment:
Nit: The most common case will be with no backslashes and this allocates empty string every time that happens. We should only do this if `numBackslash > 0`

In fact if I read it correctly, this allocates empty string for every single character on the input.